baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/SortedStreamFirstElement.java

```diff
@@ -0,0 +1,100 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import java.util.Comparator;
+import java.util.stream.Stream;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = SeverityLevel.SUGGESTION,
+        summary = "Using Stream::min is more efficient than finding the first element of the sorted stream. "
+                + "Stream::min performs a linear scan through the stream to find the smallest element.")
+public final class SortedStreamFirstElement extends BugChecker implements BugChecker.MethodInvocationTreeMatcher {
+
+    private static final Matcher<ExpressionTree> STREAM_SORTED_FIND_FIRST_MATCHER = MethodMatchers.instanceMethod()
+            .onDescendantOf(Stream.class.getName())
+            .namedAnyOf("findFirst")
+            .withNoParameters();
+
+    private static final Matcher<ExpressionTree> STREAM_SORTED_NO_PARAMS_MATCHER = MethodMatchers.instanceMethod()
+            .onDescendantOf(Stream.class.getName())
+            .named("sorted")
+            .withNoParameters();
+
+    private static final Matcher<ExpressionTree> STREAM_SORTED_WITH_COMPARATOR_MATCHER = MethodMatchers.instanceMethod()
+            .onDescendantOf(Stream.class.getName())
+            .named("sorted")
+            .withParameters(Comparator.class.getName());
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (!STREAM_SORTED_FIND_FIRST_MATCHER.matches(tree, state)) {
+            return Description.NO_MATCH;
+        }
+
+        ExpressionTree sorted = ASTHelpers.getReceiver(tree);
+        if (sorted == null) {
+            // Not expected.
+            return Description.NO_MATCH;
+        }
+        MethodInvocationTree sortedTree = (MethodInvocationTree) sorted;
+        ExpressionTree stream = ASTHelpers.getReceiver(sorted);
+        if (stream == null) {
+            // Not expected.
+            return Description.NO_MATCH;
+        }
+
+        if (Matchers.receiverOfInvocation(STREAM_SORTED_NO_PARAMS_MATCHER).matches(tree, state)) {
```

Comment:
Let's extract the `Matchers.receiverOfInvocation(` `Matcher` instances to static fields so they can be reused.